packages/gatsby/src/redux/nodes.js

```diff
@@ -154,3 +154,124 @@ const addResolvedNodes = (typeName, arr) => {
 }
 
 exports.addResolvedNodes = addResolvedNodes
+
+/**
+ * Given a ("flat") filter path leading up to "eq", a set of node types, and a
+ * cache, create a cache that for each resulting value of the filter contains
+ * all the Nodes in a Set (or, if the property is `id`, just the Nodes).
+ * This cache is used for applying the filter and is a massive improvement over
+ * looping over all the nodes, when the number of pages (/nodes) scale up.
+ *
+ * @param {Array<string>} chain
+ * @param {Array<string>} nodeTypeNames
+ * @param {undefined | Map<string, Map<string | number | boolean, Node>>} typedKeyValueIndexes
+ *   This object lives in query/query-runner.js and is passed down runQuery
+ * @returns {undefined}
+ */
+const ensureIndexByTypedChain = (
+  chain,
+  nodeTypeNames,
+  typedKeyValueIndexes
+) => {
+  const chained = chain.join(`+`)
+
+  const nodeTypeNamePrefix = nodeTypeNames.join(`,`) + `/`
+  // The format of the typedKey is `type,type/path+to+eqobj`
+  const typedKey = nodeTypeNamePrefix + chained
+
+  let byKeyValue = typedKeyValueIndexes.get(typedKey)
+  if (byKeyValue) {
+    return
+  }
+
+  const { nodes, resolvedNodesCache } = store.getState()
+
+  byKeyValue = new Map() // Map<node.value, Set<all nodes with this value for this key>>
+  typedKeyValueIndexes.set(typedKey, byKeyValue)
+
+  nodes.forEach(node => {
+    if (!nodeTypeNames.includes(node.internal.type)) {
+      return
+    }
+
+    // There can be a filter that targets `__gatsby_resolved` so fix that first
+    if (!node.__gatsby_resolved) {
+      const typeName = node.internal.type
+      const resolvedNodes = resolvedNodesCache.get(typeName)
```

Comment:
[nit] we could move that retrieval over nodes and not do it for every node

Reply:
So this would be something like this?

```js
let hash = nodeTypeNames.reduce((obj, typeName) => {
  obj[now] = resolvedNodesCache.get(typeName)
}, {})
...
node.__gatsby_resolved = hash[node.internal.type]
```

Does the overhead of refetching outweigh the overhead of running this before the main loop?